lex.go

```diff
@@ -43,14 +44,51 @@ var tokChanCap = TokenChanCap // capacity of lexer token channel
 func SetTokenChanCap(size int) {
 	tokChanCap = size
 }
-func Lex(reader io.Reader) chan NgxToken {
+
+type ExtLexer interface {
+	Register(scanner *extScanner) []string
+	Lex(matchedToken string) <-chan NgxToken
+}
+
+// LexOptions allows customization of the lexing process by specifying external lexers
+// that can handle specific directives. By registering interest in particular directives,
+// external lexers can ensure that these directives are processed separately
+// from the general lexical analysis logic.
+type LexOptions struct {
+	ExternalLexers []ExtLexer
+}
+
+func LexWithOptions(r io.Reader, options LexOptions) chan NgxToken {
 	tc := make(chan NgxToken, tokChanCap)
-	go tokenize(reader, tc)
+	go tokenize(r, tc, options)
 	return tc
 }
 
-//nolint:gocyclo,funlen,gocognit
-func tokenize(reader io.Reader, tokenCh chan NgxToken) {
+func Lex(reader io.Reader) chan NgxToken {
+	return LexWithOptions(reader, LexOptions{})
+}
+
+type extScanner struct {
```

Comment:
I think you need to make this an exported type with a better name as it's used as an argument in an exported interface. Maybe we also need to think about how else one might use this type and how it is constructed.

Reply:
done, change it to SubScanner